plugins/outputs/postgresql/tables/table_manager.go

```diff
@@ -0,0 +1,245 @@
+// this function will add the new columns with the required data type.
+func (t *defTableManager) AddColumnsToTable(db db.Wrapper, tableName string, columnIndices []int, colDetails *utils.TargetColumns) error {
+	if db == nil {
+		return errors.New("database connection is nil")
+	}
+	if tableName == "" || columnIndices == nil || colDetails == nil || len(columnIndices) == 0 {
+		errStr := fmt.Sprintf("attempted to add new columns to table '%s'. indices: %v, details: %v", tableName, columnIndices, colDetails)
+		return fmt.Errorf(errStr)
+	}
+	fullTableName := utils.FullTableName(t.schema, tableName).Sanitize()
+	for _, colIndex := range columnIndices {
+		name := colDetails.Names[colIndex]
+		dataType := colDetails.DataTypes[colIndex]
+		addColumnQuery := fmt.Sprintf(addColumnTemplate, fullTableName, utils.QuoteIdent(name), dataType)
+		_, err := db.Exec(addColumnQuery)
+		if err != nil {
+			return fmt.Errorf(
+				"E! Couldn't add missing columns to the table: %s\nError executing: %s\n%v",
+				tableName, addColumnQuery, err)
+		}
+	}
+
+	return nil
+}
+
+// Populate the 'tableTemplate' (supplied as config option to the plugin) with the details of
+// the required columns for the measurement to create a 'CREATE TABLE' SQL statement.
+// The order, column names and data types are given in 'colDetails'.
+func (t *defTableManager) generateCreateTableSQL(tableName string, colDetails *utils.TargetColumns) string {
+	colDefs := make([]string, len(colDetails.Names))
+	var pk []string
+	for colIndex, colName := range colDetails.Names {
+		colDefs[colIndex] = utils.QuoteIdent(colName) + " " + string(colDetails.DataTypes[colIndex])
+		if colDetails.Roles[colIndex] != utils.FieldColType {
+			pk = append(pk, colName)
+		}
+	}
+
+	fullTableName := utils.FullTableName(t.schema, tableName).Sanitize()
+	query := fmt.Sprintf(metricTableSQLTemplate, fullTableName, strings.Join(colDefs, ","))
+	if t.createHypertables {
+		query = query + ";" + fmt.Sprintf(createHypertableSQLTemplate, utils.QuoteLiteral(fullTableName), t.chunkTimeInterval)
+	}
+	return query
+}
+
+func (t *defTableManager) generateCreateTagTableSQL(tableName string, colDetails *utils.TargetColumns) string {
+	colDefs := make([]string, len(colDetails.Names))
+	var pk []string
+	for colIndex, colName := range colDetails.Names {
+		colDefs[colIndex] = utils.QuoteIdent(colName) + " " + string(colDetails.DataTypes[colIndex])
+		if colDetails.Roles[colIndex] != utils.FieldColType {
+			pk = append(pk, colName)
+		}
+	}
+
+	fullTableName := utils.FullTableName(t.schema, tableName).Sanitize()
+	return fmt.Sprintf(tagTableSQLTemplate, fullTableName, strings.Join(colDefs, ","))
+}
+
+// For a given table and an array of column names it checks the database if those columns exist,
+// and what's their data type.
+func (t *defTableManager) findColumnPresence(db db.Wrapper, tableName string, columns []string) ([]*columnInDbDef, error) {
```

Comment:
I haven't checked if this is possible, but can we just attempt to alter all columns on the table?  This way we don't need to know what is in the database and it removes a race condition.